controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -408,30 +493,41 @@ bool master_thread::autoconfig_wsc_add_m2(std::shared_ptr<ieee1905_1::tlvWscM1>
             ? WSC::FRONTHAUL_BSS
             : (bss_type & WSC::BACKHAUL_BSS ? WSC::BACKHAUL_BSS : WSC::TEARDOWN);
     tlvWscM2->primary_device_type_attr().sub_category_id = WSC::WSC_DEV_NETWORK_INFRA_GATEWAY;
-    // Encrypted settings
-    auto encrypted_settings = tlvWscM2->create_encrypted_settings();
-    if (!encrypted_settings) {
-        LOG(ERROR) << "Failed to create encrypted settings";
+
+    ///////////////////////////////
+    // @brief encryption support //
+    ///////////////////////////////
+    if (!autoconfig_wsc_calculate_keys(tlvWscM1, tlvWscM2))
         return false;
-    }
-    // TODO - actually encrypt the data
-    encrypted_settings->set_ssid("prplMesh_ssid");
-    encrypted_settings->authentication_type_attr().data = WSC::WSC_AUTH_WPA2; //DUMMY
-    encrypted_settings->encryption_type_attr().data     = WSC::WSC_ENCR_AES;
-    std::fill(encrypted_settings->network_key_attr().data,
-              encrypted_settings->network_key_attr().data +
-                  encrypted_settings->network_key_attr().data_length,
+
+    // Encrypted settings
+    // Encrypted settings are the ConfigData + IV. First create the ConfigData,
+    // Then copy it to the encrypted data, add an IV and encrypt.
+    // Finally, add HMAC
+
+    // Create ConfigData
+    uint8_t buf[1024];
+    WSC::cConfigData config_data(buf, sizeof(buf), false, true);
+    config_data.set_ssid("prplMesh-ssid");
+    config_data.authentication_type_attr().data = WSC::WSC_AUTH_WPA2;
+    config_data.encryption_type_attr().data     = WSC::WSC_ENCR_AES;
+    std::fill(config_data.network_key_attr().data,
+              config_data.network_key_attr().data + config_data.network_key_attr().data_length,
               0xaa); //DUMMY
-    encrypted_settings->bssid_attr().data = tlvWscM1->mac_attr().data;
-    std::fill(encrypted_settings->key_wrap_auth_attr().data,
-              encrypted_settings->key_wrap_auth_attr().data +
-                  encrypted_settings->key_wrap_auth_attr().data_length,
+    std::fill(config_data.key_wrap_auth_attr().data,
+              config_data.key_wrap_auth_attr().data + config_data.key_wrap_auth_attr().data_length,
               0xff); //DUMMY
 
-    if (!tlvWscM2->add_encrypted_settings(encrypted_settings)) {
-        LOG(ERROR) << "Failed to add encrypted settings";
+    // TODO - remove this print before merging!
```

Comment:
Don't forget :-)

Reply:
wrapping in `#ifdef DEBUG` so we have it in debug builds, ok?

Reply:
> wrapping in `#ifdef DEBUG` so we have it in debug builds, ok?

Then you can just leave it in regardless. It's not as if any secrets are printed here. And anyway, if there were, they would be known in the controller already.